pkg/quota/evaluator/core/persistent_volume_claims.go

```diff
@@ -23,63 +23,124 @@ import (
 	"k8s.io/kubernetes/pkg/quota"
 	"k8s.io/kubernetes/pkg/quota/generic"
 	"k8s.io/kubernetes/pkg/runtime"
 	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/watch"
 )
 
+const storageClassSuffix string = ".storage-class.kubernetes.io/"
+
 // NewPersistentVolumeClaimEvaluator returns an evaluator that can evaluate persistent volume claims
 func NewPersistentVolumeClaimEvaluator(kubeClient clientset.Interface) quota.Evaluator {
+	storageClassStore := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+				return kubeClient.Storage().StorageClasses().List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return kubeClient.Storage().StorageClasses().Watch(options)
+			},
+		},
+		&storage.StorageClass{},
+		storageClassStore,
+		0,
+	)
+	reflector.Run()
+
+	matchesResourceNamesFunc := makeMatchedResourceNamesFunc(storageClassStore)
+	usageFunc := makePersistentVolumeClaimUsageFunc(storageClassStore, matchesResourceNamesFunc)
+	constraintsFunc := makePersistentVolumeClaimConstraintsFunc(usageFunc)
 	allResources := []api.ResourceName{api.ResourcePersistentVolumeClaims, api.ResourceRequestsStorage}
 	return &generic.GenericEvaluator{
 		Name:              "Evaluator.PersistentVolumeClaim",
 		InternalGroupKind: api.Kind("PersistentVolumeClaim"),
 		InternalOperationResources: map[admission.Operation][]api.ResourceName{
 			admission.Create: allResources,
 		},
-		MatchedResourceNames: allResources,
-		MatchesScopeFunc:     generic.MatchesNoScopeFunc,
-		ConstraintsFunc:      PersistentVolumeClaimConstraintsFunc,
-		UsageFunc:            PersistentVolumeClaimUsageFunc,
+		MatchedResourceNames:     allResources,
+		MatchedResourceNamesFunc: matchesResourceNamesFunc,
+		MatchesScopeFunc:         generic.MatchesNoScopeFunc,
+		ConstraintsFunc:          constraintsFunc,
+		UsageFunc:                usageFunc,
 		ListFuncByNamespace: func(namespace string, options api.ListOptions) (runtime.Object, error) {
 			return kubeClient.Core().PersistentVolumeClaims(namespace).List(options)
 		},
 	}
 }
 
-// PersistentVolumeClaimUsageFunc knows how to measure usage associated with persistent volume claims
-func PersistentVolumeClaimUsageFunc(object runtime.Object) api.ResourceList {
-	pvc, ok := object.(*api.PersistentVolumeClaim)
-	if !ok {
-		return api.ResourceList{}
-	}
-	result := api.ResourceList{}
-	result[api.ResourcePersistentVolumeClaims] = resource.MustParse("1")
-	if request, found := pvc.Spec.Resources.Requests[api.ResourceStorage]; found {
-		result[api.ResourceRequestsStorage] = request
+// makeMatchedResourceNamesFunc returns a function that returns the list of resources matched
+func makeMatchedResourceNamesFunc(storageClassStore cache.Store) generic.MatchesResourceNamesFunc {
+	allResources := []api.ResourceName{api.ResourcePersistentVolumeClaims, api.ResourceRequestsStorage}
```

Comment:
pretty much, the idea was to support splitting all pvc related quota items per storage class if desired.